Products/CMFPlomino/PlominoReplicationManager.py

```diff
@@ -1304,6 +1329,82 @@ def saveFile(path, content):
+        fieldnames = range(max_columns)
+        column_number_field_name_mapping = {0: "doc_id"}
+
+        for form in forms:
+            for field in form.getFormFields():
+                if field.id not in column_number_field_name_mapping.itervalues():
+                    column_number_field_name_mapping[len(column_number_field_name_mapping)] = field.id
+
+        if docids:
+            doc_ids = [self.getDocument(i) for i in docids]
+        else:
+            doc_ids = self.documents.keys()
+
+        number_of_docs = len(doc_ids)
+        logger.info("Exporting %s documents..." % number_of_docs)
+
+        # Returns a tuple of ({doc_data}, [new_field_names]) 
+        def _iterate_documents(doc_ids_list):
+            for i, doc_id in enumerate(doc_ids_list):
+                if i % 2000 == 0:
+                    transaction.abort()
+                yield self.documents[doc_id]
+
+        transaction.abort()
+        logger.info("All fieldnames checked. Starting document writing")
+
+        zope_export_folder_path = getattr(getConfiguration(), 'clienthome', "")
+        export_folder_path = os.path.join(zope_export_folder_path, 'export', self.id)
+        export_path = os.path.join(export_folder_path, '%s.csv' % self.id)
+        if os.path.isdir(export_folder_path):
+            # remove previous export
+            if os.path.exists(export_path):
+                os.remove(export_path)
+        else:
+            os.makedirs(export_folder_path)
+
+        with TemporaryFile(mode="w+b") as csv_data_tempfile:
+            writer = csv.DictWriter(csv_data_tempfile, fieldnames=fieldnames)
+
+            transaction.abort() 
+
+            for i, doc in enumerate(_iterate_documents(doc_ids)):
+                for field_name in doc.getItems():
+                    if field_name not in column_number_field_name_mapping.itervalues():
+                        print("Adding field %s" % field_name)
+                        column_number_field_name_mapping[len(column_number_field_name_mapping)] = field_name
+
+                document_values = {0: doc.id}
+                for column_number in column_number_field_name_mapping:
+                    field_name = column_number_field_name_mapping[column_number]
+                    document_values[column_number] = doc.getItem(field_name, "") or ""
+
+                empty_columns_to_add = max_columns - len(document_values) - 1
+                for column_number in range(empty_columns_to_add, len(document_values) + empty_columns_to_add):
+                    document_values[column_number] = ""
+
+                if i % 20000 == 0:
+                    print("Documents exported: %s/%s" % (i, number_of_docs))
+                    logger.info("Documents exported: %s/%s" % (i, number_of_docs))
+
+                row = safe_dict(document_values)
+                writer.writerow(row)
+
+            transaction.abort()
+            logger.info("All documents written. Closing CSV.")
+
```

Comment:
where are you writing out the header? and is there a plan to trim the extra cols?